### ISSUE TITLE
Razer Command Dial (on Razer BlackWidow v4 Pro) doesn't work in Driver Mode

### ISSUE THREAD

**Issue by therob89** [openrazer/razer-drivers#2228]
### Distribution

OS: Fedora 39
Kernel: 6.7.10-200.fc39.x86_64
OpenRazer: 3.7.0-1.3

### Support

- [x] Have you tried rebooting?
- [x] Have you looked at the [Troubleshooting](https://github.com/openrazer/openrazer/wiki/Troubleshooting) page?

Razer Command Dial on Razer BlackWidow4 Pro doesn't work in **driver mode**.

<img width="842" alt="image" src="https://github.com/openrazer/openrazer/assets/8765735/5ea6fb62-a76d-46e8-85e7-b18b6c4c21ec">

 A partial workaround to this problem was the proposed by @z3ntu that works until the next boot:

1. Identify the BlackWidow v4 Pro Device

  <img width="481" alt="image" src="https://github.com/openrazer/openrazer/assets/8765735/779c19b2-7830-42e6-b0ab-a01d01b28ab4">
  
  ```bash
  # Identify the Razer BlackWidow v4 Pro Device
  find /sys -name device_mode
  hexdump /sys/devices/pci0000:00/0000:00:14.0/usb3/3-7/3-7.3/3-7.3.1/3-7.3.1:1.2/0003:1532:028D.0007/device_mode
  ```
 
2. Switch to Device Mode

  ```bash
  # Identify the Razer BlackWidow v4 Pro Device
  find /sys -name device_mode
  hexdump /sys/devices/pci0000:00/0000:00:14.0/usb3/3-7/3-7.3/3-7.3.1/3-7.3.1:1.2/0003:1532:028D.0007/device_mode


  # Switch the Device in Device Mode to use Razer Command Dial
  echo -n -e "\x00\x00" > /sys/devices/pci0000:00/0000:00:14.0/usb3/3-7/3-7.3/3-7.3.1/3-7.3.1:1.2/0003:1532:028D.0007/device_mode
  ```
<img width="482" alt="image" src="https://github.com/openrazer/openrazer/assets/8765735/bf702fb7-853e-4a66-b372-8c2bc9fd4637">

Below the LoC where the Driver mode is set:

https://github.com/openrazer/openrazer/blob/master/daemon/openrazer_daemon/hardware/keyboards.py#L31

**Comment by z3ntu** [comment]
Would be interesting if you could add some logging into https://github.com/openrazer/openrazer/blob/479a8f330bf126abee6c54ca9ccfd3af7f4e2ed3/driver/razerkbd_driver.c#L3230 and/or https://github.com/openrazer/openrazer/blob/479a8f330bf126abee6c54ca9ccfd3af7f4e2ed3/driver/razerkbd_driver.c#L3578 to see what kind of data the device sends us. Then we can try and handle this to work as expected also in driver mode. If you need any more specific advice let me know.

See also https://github.com/openrazer/openrazer/blob/master/DEVELOPMENT.md

**Comment by Edu4rdSHL** [comment]
Hi, I'm interested in getting it working, however, the quoted lines in the previous comment seem to point to something different. Can you please update them?

Also, what would be the exact procedure? Add logging, press the key, report the log back?

Thanks!

**Comment by z3ntu** [comment]
@Edu4rdSHL edited the previous comment

and yes, add logging there, do the things to reverse engineer, take notes what's being printed for what, and report back that.

**Comment by Edu4rdSHL** [comment]
Ping @z3ntu

**Comment by z3ntu** [comment]
@Edu4rdSHL Someone with the device will need to add the logging and figure out what needs to be done. I can't do anything without the device.

**Comment by Edu4rdSHL** [comment]
@z3ntu sorry, I confused this issue with the polling rate one while I was half asleep. My bad.

**Comment by thedustinwidmann-lang** [comment]
I have a vested interest in seeing this functionality improved, so I've done my best to capture the data that @z3ntu asked for.

I added a little bit of code to razerkbd_driver.c

```c
static int razer_raw_event_bitfield(struct hid_device *hdev, struct razer_kbd_device *device, struct usb_interface *intf, struct hid_report *report, u8 *data, int size)
{
    pr_info("[RAZERKBD] >>> ENTER razer_raw_event_bitfield <<<\n");

    int idbg;

    dev_info(&hdev->dev, "[RAZERKBD] raw_event: report_id=0x%02x size=%d\n", report->id, size);
    
    for (idbg = 0; idbg < size; idbg++)
    {
        pr_info("[RAZERKBD] raw_event data[%02d] = 0x%02x\n", idbg, data[idbg]);
    }   
```

and then took notes on as many key actions as I could. Please see the attached files.

[keys.csv](https://github.com/user-attachments/files/23830392/keys.csv)

[evtest.txt](https://github.com/user-attachments/files/23830396/evtest.txt)

**Comment by z3ntu** [comment]
@thedustinwidmann-lang Thanks for sharing this! However I'm not quite sure how to interpret the data, maybe I'm also not quite aware what the device does.

You have this dial on the top left that you can press, turn left and turn right. Why do you have different colors along with the command dial logs (cdial green cw, cdial red ccw, etc)?

And just to confirm (sorry that it's a bit much later), did you check that the device is indeed in driver mode when you captured the logs?

**Comment by ShaiAvr** [comment]
@z3ntu I am dual booting Windows and CachyOS, and I can explain how the device works.

> You have this dial on the top left that you can press, turn left and turn right. Why do you have different colors along with the command dial logs (cdial green cw, cdial red ccw, etc)?

The dial allows to configure several actions that can be performed with it. Pressing the dial changes the action it would perform, and its color to indicate that action. On Windows, it can be configured in Razer Synapse:

<img width="1444" height="1112" alt="Image" src="https://github.com/user-attachments/assets/fbadd17e-8f7f-4b30-8048-b264bf1419e9" />

As you can see, there are several built-in actions like scrolling horizontically/vertically, zooming windows, switching apps (Alt tabbing), etc, and I can also create custom actions. In the screenshot above, you can see that pressing the dial cycles between the colors pink, white, red and orange, and cycles between their corresponding actions. Once the dial has the color of the action I want, I can turn it to perform the corresponding action. For example, I press it multiple times until it turns pink, and then turning it would scroll vertically, like a mouse wheel. Pressing it again would change its color to white and then turning it would allow me to scroll horizontally (shift + mouse wheel).

On Windows, the dial functions as expected, but on CachyOS, it doesn't do anything.

**Comment by thedustinwidmann-lang** [comment]
@z3ntu Yes, it was in device mode when I did that capture, at least as near as I can tell? It was behaving the way it does on Windows, more or less. (so programmable hotkeys and the like weren't working). The description @ShaiAvr gave more or less covers the functionality, though I didn't really track the additional features that can be configured (I've only used this in Linux since I pulled it out of the box). I guess I could do that when I get some free time.  If you can steer the boat I wouldn't mind taking a swing at making some edits?

**Comment by thedustinwidmann-lang** [comment]
@z3ntu Following up with the data you asked for, this time captured in
confirmed **driver mode** (`device_mode` = `03 00`, stock 3.12.2, raw USB
interrupt streams via `usbhid-dump`, verified with evtest end-to-end).

**Short version: the driver already delivers everything.** Dial press
arrives as `04 60` on the keyboard interface and is translated to
`KEY_F24` (since v3.7.0, the first release supporting this device);
rotation arrives on the mouse interface as Consumer AC Pan and plain
hid-input emits `REL_HWHEEL ±1` per click. The dial only *feels* dead
because nothing consumes those events by default. There is no on-device
mode logic in driver mode (rotation reports are identical across presses
and the dial LED never changes) — the mode/color behavior seen on Windows
is Synapse-side policy, so a Linux equivalent is userspace work, which
fits this issue's labels.

Attached:

- **protocol-reference.md** — complete driver-mode protocol: interfaces,
  report layouts, descriptor decode, driver code paths, and notes for
  users landing here.
- **supplement-to-previous-captures.md** — corrections/additions to my
  2025-11 attachments, which were taken in device mode and answer a
  different question.

I don't believe any kernel or daemon change is needed. For users wanting
Synapse-style modes today, I've built a userspace daemon around these
findings (press cycles modes, per-mode rotation actions, dial LED color):
https://gitlab.com/dustinwidmann/detentd — I think this issue can be closed?

[supplement-to-previous-captures.md](https://github.com/user-attachments/files/29651856/supplement-to-previous-captures.md)

[protocol-reference.md](https://github.com/user-attachments/files/29651868/protocol-reference.md)
